### ISSUE TITLE
[CRITICAL] Clair Scannner 400 errors when Analyzing code

### ISSUE THREAD

**Issue by vbodla01** [arminc/clair-scanner#141]
Running clair-scanner in docker and we are noticing `blobs/sha256` appended to the layer and the scanning is failing.
```
23:26:52  + /clair/clair-scanner --clair=http://clair/ --ip=X.X.X.X --report=build/clair/clair-result_1.json --whitelist=build/clair/clair-allowlist.yaml sha256:eb9d02bc6996d6ff8e8b20e6ab12d97ae9288e1868461e640d6caa6f17acc754
23:26:52  2024/12/09 04:26:51 [0;32m[INFO] ▶ Start clair-scanner[0m
23:27:00  2024/12/09 04:26:59 [0;32m[INFO] ▶ Server listening on port 9279[0m
23:27:00  2024/12/09 04:26:59 [0;32m[INFO] ▶ Analyzing blobs/sha256/a9b34f345c5b4d6b4e95067e09312cad6db11e6593eb93dd73df573dff9b3f02[0m
23:27:00  2024/12/09 04:26:59 [0;31m[CRIT] ▶ Could not analyze layer: Clair responded with a failure: Got response 400 with message {"Error":{"Message":"could not find layer"}}
23:27:00  [0m
23:27:00  [Pipeline] echo
23:27:00  Clair returned with status 1
23:27:00  [Pipeline] archiveArtifacts
```
Earlier scans didn't have `blobs/sha256` infront of the layer.
Tried pointing to older image for clair scanner but that didn't work either.

**Comment by vbodla01** [comment]
@arminc  - Could you please assist me on this, we are running this on kube pod and not able to go further.

**Comment by vbodla01** [comment]
Thank you @arminc for the update.
Please let me know once the initial release is available or if you have any band-aid for this.

**Comment by vbodla01** [comment]
@arminc - Following up to see if there is any update on the fix.

**Comment by vbodla01** [comment]
@arminc - Do you have any update on this issue ?

**Comment by arminc** [comment]
@vbodla01 I am currently not actively working on this project, @datadot might know more.

**Comment by vbodla01** [comment]
Thank you @arminc.
I'll check with @datadot going forward.

**Comment by vbodla01** [comment]
Hello @datadot 
Is there any schedule available for the fix rollout ?

**Comment by datadot** [comment]
@vbodla01 I have just made a prerelease which will hopefully help address the issues you're experiencing...

[14.0.0-prerelease-0](https://github.com/arminc/clair-scanner/releases/tag/14.0.0-prerelease-0)

This prerelease does require some updates to the whole stack and unfortunately I haven't had a chance yet to fully document everything:

1) Use arminc/clair-db-v4 instead of arminc/clair-db
2) Use arminc/clair-local-scan-v4 instead of arminc/clair-local-scan

Please let me know how you get along - I'll be running more extensive tests over the coming days before making formal release.

**Comment by vbodla01** [comment]
Thank you @datadot 
I updated deployment to get images from v4 and tried a scan and it is failing with below errors,
```
04:42:15  2025/01/08 09:42:15 [0;32m[INFO] ▶ Start clair-scanner[0m
04:43:02  2025/01/08 09:42:59 [0;32m[INFO] ▶ Server listening on port 9279[0m
04:43:02  2025/01/08 09:42:59 [0;32m[INFO] ▶ Analyzing blobs/sha256/81a2e9bd12d61418ce224142c577ed20d3c7ae5083e34dfb7f14bfee9bef9de8[0m
04:43:02  2025/01/08 09:42:59 [0;31m[CRIT] ▶ Could not analyze layer: POST to Clair failed Post http://clair/v1/layers: dial tcp 172.21.35.249:80: getsockopt: connection refused[0m
```

**Comment by vbodla01** [comment]
@datadot -  Did you get a chance to look at the new error.

**Comment by datadot** [comment]
@vbodla01 - apologies only just had chance to take a look at this.  The log lines you have pasted above look to be from the old v12 cli tool.  New logs should be along the lines of:

> ./clair-scanner_linux-arm64 --clair=http://localhost:8080 nginx:1.16.0
> 2025/01/09 19:53:39 [INFO] ▶ Starting clair-scanner for image: nginx:1.16.0
> 2025/01/09 19:53:41 [INFO] ▶ Starting HTTP file server on 0.0.0.0:9279 serving path: /tmp/clair-scanner-51587554
> 2025/01/09 19:53:41 [INFO] ▶ Starting HTTP server on 0.0.0.0:9279
> 2025/01/09 19:53:41 [INFO] ▶ HTTP server is ready on 0.0.0.0:9279
> 2025/01/09 19:53:54 [INFO] ▶ Image [nginx:1.16.0] contains NO unapproved vulnerabilities
> 2025/01/09 19:53:54 [INFO] ▶ Scan completed successfully with no vulnerabilities

Would you mind checking the tool you're using?

[14.0.0-prerelease-0](https://github.com/arminc/clair-scanner/releases/tag/14.0.0-prerelease-0)

Thanks

**Comment by vbodla01** [comment]
Thank you @datadot  for your response.

I've used latest image tags and running them in Kubernetes
https://hub.docker.com/r/arminc/clair-db-v4/tags
https://hub.docker.com/r/arminc/clair-local-scan-v4/tags

**Comment by datadot** [comment]
Hey @vbodla01 that shouldn't be a problem.  It's just the log output you provided looks like the logs from v12, the updated tool generates different logs.  E.g: "Starting clair-scanner for image: nginx:1.16.0" not "Start clair-scanner".

**Comment by vbodla01** [comment]
Thank you @datadot 
I was able to fix our things internally and this is what I got now,

```
14:21:14  2025/01/14 19:21:14 [0;32m[INFO] ▶ Starting clair-scanner for image: docker-eu.artifactory.swg-devops.com/sys-zaas-team-iaas-dev-docker-local/zaas/scaas-manager:hm-vpc-v1.3.2[0m
14:21:47  2025/01/14 19:21:44 [0;32m[INFO] ▶ Starting HTTP file server on 0.0.0.0:9279 serving path: /tmp/clair-scanner-2869086912[0m
14:21:47  2025/01/14 19:21:44 [0;32m[INFO] ▶ Starting HTTP server on 0.0.0.0:9279[0m
14:21:47  2025/01/14 19:21:44 [0;32m[INFO] ▶ HTTP server is ready on 0.0.0.0:9279[0m
14:22:19  2025/01/14 19:22:14 Failed to submit container for analysis: error sending request: Post "http://clair/indexer/api/v1/index_report": dial tcp 172.21.37.82:80: i/o timeout
14:22:19  [Pipeline] echo
14:22:19  Clair exited with 1
```

**Comment by datadot** [comment]
Hey @vbodla01, this is looking like the port is potentially wrong.  There are multiple layers here because of it running on k8s, so apologies if I'm covering ground you already have.

1) Clair V4 container port is now 8080 not 80.
2) I'm assuming you're running this through a k8s service, so you may already have updated?
3) Can you check that you can reach the endpoint /indexer/api/v1/index_report?

Thanks.

**Comment by vbodla01** [comment]
Hello @datadot 

Yes, we are running in K8's and we changed the service to pick on 80 and route it to 6060.
I tried in local and got some similar results like I mentioned earlier.

Here are the steps I followed,
`docker run -d --name clair-db-4 arminc/clair-db-v4`
`docker run -p 6060:6060 --link clair-db-4:postgres -d --name clair arminc/clair-local-scan-v4:latest`

```
./clair-scanner_linux-amd64 --clair=http://localhost:8080 nginx:latest
2025/01/15 15:43:01 [INFO] ▶ Starting clair-scanner for image: nginx:latest
2025/01/15 15:43:05 [INFO] ▶ Starting HTTP file server on 0.0.0.0:9279 serving path: /tmp/clair-scanner-515203418
2025/01/15 15:43:05 [INFO] ▶ Starting HTTP server on 0.0.0.0:9279
2025/01/15 15:43:06 [INFO] ▶ HTTP server is ready on 0.0.0.0:9279
2025/01/15 15:43:06 Failed to submit container for analysis: error sending request: 
Post "http://localhost:8080/indexer/api/v1/index_report": dial tcp 127.0.0.1:8080: connect: connection refused
```


Started with Port 8080 as you mentioned,
`docker run  -p 8080:8080 --link clair-db-4:postgres -d --name clair arminc/clair-local-scan-v4:latest`
`docker run -d --name clair-db-4 arminc/clair-db-v4`

Output:
```
./clair-scanner_linux-amd64 --clair=http://localhost:8080 nginx:latest
2025/01/15 15:54:22 [INFO] ▶ Starting clair-scanner for image: nginx:latest
2025/01/15 15:54:24 [INFO] ▶ Starting HTTP file server on 0.0.0.0:9279 serving path: /tmp/clair-scanner-3298042022
2025/01/15 15:54:24 [INFO] ▶ Starting HTTP server on 0.0.0.0:9279
2025/01/15 15:54:24 [INFO] ▶ HTTP server is ready on 0.0.0.0:9279
2025/01/15 15:54:24 Failed to submit container for analysis: error parsing report ID: manifest_hash not found in response
```

Should I change the port 6060 to 8080 in our K8;s ?

**Comment by datadot** [comment]
Hey @vbodla01 apologies I wanted to make sure I didn't discount anything so I needed to find an x86 machine.

I think what's happening here is the uploaded layers to clair do not have the correct client IP of the cli tool that is temporarily hosting the container layers on the HTTP server that is spun up on port 9279.  Could you try your last test as follows:

./clair-scanner_linux-amd64 --clair=http://localhost:8080 --ip "$(ifconfig eth0 | awk '/inet /{print substr($2,0)}')" nginx:latest

I'm making some assumptions about the interface being eth0, etc... but if my simple IP calc doesn't work, replace it with the interface IP.

Thanks.

**Comment by vbodla01** [comment]
Hello @datadot 
We were using `hostname -i` for the ip address as we are running in the container.
I updated to what you have provided and it throws me an error,
`No such property: eth0`

**Comment by vbodla01** [comment]
Greetings @datadot 
Did you get a chance to look at the comments I updated? Do you have any updates for me?

**Comment by datadot** [comment]
Hey @vbodla01, sorry I've not had any luck in reproducing your behaviour.

Are you able to share the k8s manifests or helm chart you're using to deploy so I can spin this up in a cluster here.  Can you also verify where you're running the CLI from and if you're running any port forwards or anything between that client and the k8s services.

Thanks.

**Comment by vbodla01** [comment]
Hello @datadot 

```
node ('prod_image_with_docker') {
    stage ('Setup Clair') {
        clairyaml = '''
apiVersion: apps/v1
kind: Deployment
metadata:
  name: clair-deployment
  labels:
    app: clair
spec:
  replicas: 1
  selector:
    matchLabels:
      app: clair
  template:
    metadata:
      labels:
        app: clair
      annotations:
        refresh: label2
    spec:
      hostAliases:
      - ip: "127.0.0.1"
        hostnames:
        - "postgres"
      containers:
      - name: clair-db
        image: arminc/clair-db-v4:latest
        resources:
          requests:
            memory: "128Mi"
            cpu: "250m"
          limits:
            memory: "512Mi"
            cpu: 1
      - name: clair
        image: arminc/clair-local-scan-v4:latest
        ports:
        - containerPort: 8080
        resources:
          requests:
            memory: "128Mi"
            cpu: "250m"
          limits:
            memory: "1Gi"
            cpu: "2"

---
kind: Service
apiVersion: v1
metadata:
  name: clair
spec:
  selector:
    app: clair
  ports:
  - protocol: TCP
    port: 80
    targetPort: 8080
'''

//    # kubectl delete -f clair.yaml
        writeFile file: "clair.yaml", text: clairyaml
        sh '''
            kubectl get deployments
            kubectl delete -f clair.yaml
            kubectl apply -f clair.yaml
            kubectl rollout status deployment clair-deployment
            kubectl describe deployment clair-deployment
            kubectl describe svc clair
        '''
    }
}
```
This is what configured in our Jenkins pipeline and we are using Ubuntu 24.10.

**Comment by datadot** [comment]
Thanks @vbodla01 I'm assuming that you're running your Jenkins environment within this k8s cluster then?

**Comment by vbodla01** [comment]
Hello @datadot 
Yes, you are right.

**Comment by vbodla01** [comment]
Hello @datadot 
Following up to see if you have any updates for us?

**Comment by datadot** [comment]
Hey @vbodla01, apologies I'm waiting for some network ACLs to be modified so I can run the tests fully, however I did achieve the some error messages out of CLAIR of which one did match yours.

1) With your setup even with a relatively small container (nginx in my case) it caused an oomkill with your 1Gi limit, if you can I'd suggest increasing the limit to 4Gi to be safe.

2) Can you search your logs from the clair container for something like:

{"level":"debug","component":"httptransport/New","request_id":"a4390b4d708b0564","code":500,"error":"failed to start scan: failed to fetch layers: fetcher: encountered errors: error realizing layer sha256:d2134b0be91e9e293bc872b719538440e5f933e9828cd96430c85d904afb5aa9: fetcher: request failed: Get \"http://localhost:9279/blobs/sha256/d2134b0be91e9e293bc872b719538440e5f933e9828cd96430c85d904afb5aa9\": dial tcp [::1]:9279: connect: connection refused","time":"2025-01-23T20:24:29Z","message":"http error response"}
{"level":"info","component":"httptransport/New","request_id":"a4390b4d708b0564","remote_addr":"192.168.194.0:57389","method":"POST","request_uri":"/indexer/api/v1/index_report","status":500,"duration":357.738574,"time":"2025-01-23T20:24:29Z","message":"handled HTTP request"}

This is my latest one and is because the clair container can't access my client that I'm running the CLI from.  Once I have the network ACL updated I'll continue running the tests.

Thanks.

**Comment by vbodla01** [comment]
Hello @datadot 

I think I've good news, increasing memory to 4 Gigs solved the problem.
I was able to scan image successfully.

```
05:49:34  + /clair/clair-scanner --clair=http://clair/ --ip=X.X.X.X --report=build/clair/clair-result_docker-eu-dev-PR-43-2.json --whitelist=build/clair/clair-allowlist.yaml docker-eu-dev-PR-43-2
05:49:34  2025/01/24 10:49:34 [0;32m[INFO] ▶ Starting clair-scanner for image: hm-vpc-test-runner:dev-PR-43-2[0m
05:49:57  2025/01/24 10:49:56 [0;32m[INFO] ▶ Starting HTTP file server on 0.0.0.0:9279 serving path: /tmp/clair-scanner-3006509694[0m
05:49:57  2025/01/24 10:49:56 [0;32m[INFO] ▶ Starting HTTP server on 0.0.0.0:9279[0m
05:49:57  2025/01/24 10:49:56 [0;32m[INFO] ▶ HTTP server is ready on 0.0.0.0:9279[0m
05:50:35  2025/01/24 10:50:32 [0;33m[WARN] ▶ Image [hm-vpc-test-runner:dev-PR-43-2] contains 193 total vulnerabilities[0m
05:50:35  2025/01/24 10:50:32 [0;31m[ERRO] ▶ Image [hm-vpc-test-runner:dev-PR-43-2] contains 193 unapproved vulnerabilities[0m
05:50:35  2025/01/24 10:50:33 [0;31m[ERRO] ▶ Unapproved vulnerabilities found in the image[0m
05:50:35  [Pipeline] echo
05:50:35  Clair exited with 1
05:50:35  [Pipeline] archiveArtifacts
05:50:36  Archiving artifacts
05:50:36  [Pipeline] echo
05:50:36  Analyzing results for hm-vpc-test-runner:dev-PR-43-2
05:50:36  [Pipeline] readJSON
05:50:37  [Pipeline] writeFile
05:50:37  [Pipeline] archiveArtifacts
05:50:37  Archiving artifacts
05:50:37  [Pipeline] echo
05:50:37  Fix patches for vulnerabities found (12)
05:50:37  [Pipeline] echo
05:50:37  Vulnerabilities were found that don't have fixes available yet.
```

**Comment by vbodla01** [comment]
Hey @datadot 
Will there be an official fix released? if so, could you please provide an ETA for it.
Also, what limits do you suggest for clair-db, we are limiting to 512Mi but wanted to know the best settings.

Also, how often should we build our clair deployment to pull the `latest` images which have the updated vulnerabilities 

Thanks 
Vijay

**Comment by vbodla01** [comment]
Hello @datadot 

Following up to see if you have any updates on my comments after our previous conversation. Any ETA on the official fix?
We are currently blocked on our production releases as we are dependent on clair scans.

Also,
Do we have any output that can be printed when clair is scanning the layers like the earlier versions?

Thanks,
Vijay

**Comment by vbodla01** [comment]
Hello @datadot 
Apologies for bugging you continuously.
Did you get a chance to look at the comments I posted?

I also suspect the scan is reporting vulnerabilities improperly though we updated them.

Thanks,
Vijay

**Comment by datadot** [comment]
@vbodla01 14.0.0 will only work with v4 images.

re reporting improperly please raise this as a new issue if you believe CLAIR is returning the correct results and we are interpreting them incorrectly.
